felix/bpf/polprog/pol_prog_builder_test.go

```diff
@@ -88,15 +90,150 @@ func TestLogActionIgnored(t *testing.T) {
+	checkLabelsAndComments(genInsns("dst_net"), -1, "If dest not in {13.0.0.0/8,14.0.0.0/8}, skip to next rule", "comment", 0)
+	checkLabelsAndComments(genInsns("not_dst_net"), -1, "If dest in {12.0.0.0/8,15.0.0.0/8}, skip to next rule", "comment", 0)
+	checkLabelsAndComments(genInsns("src_ports"), -1, "If source port is not within any of {80,8080-8081}, skip to next rule", "comment", 0)
+	checkLabelsAndComments(genInsns("dst_ports"), -1, "If dest port is not within any of {90-95,9090-9091}, skip to next rule", "comment", 0)
+	checkLabelsAndComments(genInsns("not_src_ports"), -1, "If source port is within any of {80,8080-8081}, skip to next rule", "comment", 0)
+	checkLabelsAndComments(genInsns("not_dst_ports"), -1, "If dest port is within any of {90-95,9090-9091}, skip to next rule", "comment", 0)
+
+	checkLabelsAndComments(genInsns("src_named_port"), -1, "If source port is not within any of the named ports {n:abcdef1234567890}, skip to next rule", "comment", 1)
+	checkLabelsAndComments(genInsns("not_src_named_port"), -1, "If source port is within any of the named ports {n:abcdef1234567890}, skip to next rule", "comment", 1)
+	checkLabelsAndComments(genInsns("dst_named_port"), -1, "If dest port is not within any of the named ports {n:foo1234567890}, skip to next rule", "comment", 1)
+	checkLabelsAndComments(genInsns("not_dst_named_port"), -1, "If dest port is within any of the named ports {n:foo1234567890}, skip to next rule", "comment", 1)
+
+	checkLabelsAndComments(genInsns("src_ipset_ids"), -1, "If source doesn't match ipset s:sbcdef1234567890, skip to next rule", "comment", 0)
+	checkLabelsAndComments(genInsns("not_src_ipset_ids"), -1, "If source matches ipset s:sbcdef1234567890, skip to next rule", "comment", 0)
+	checkLabelsAndComments(genInsns("dst_ipset_ids"), -1, "If dest doesn't match ipset d:sbcdef1234567890, skip to next rule", "comment", 0)
+	checkLabelsAndComments(genInsns("not_dst_ipset_ids"), -1, "If dest matches ipset d:sbcdef1234567890, skip to next rule", "comment", 0)
+
+	checkLabelsAndComments(genInsns("icmp_type_code"), -1, "If ICMP type != 10 or code != 12, skip to next rule", "comment", 0)
+	checkLabelsAndComments(genInsns("not_icmp_type_code"), -1, "If ICMP type == 10 and code == 12, skip to next rule", "comment", 0)
+	checkLabelsAndComments(genInsns("icmp_type"), -1, "If ICMP type != 10, skip to next rule", "comment", 0)
+	checkLabelsAndComments(genInsns("not_icmp_type"), -1, "If ICMP type == 10, skip to next rule", "comment", 0)
+}
+
+func checkIfStringExists(strArray []string, expectedStr string) bool {
+	for _, str := range strArray {
+		if str == expectedStr {
+			return true
+		}
+	}
+	return false
+}
+
+func getStartofTierInsnIndex(insns asm.Insns) int {
+	for i, in := range insns {
+		for _, comment := range in.Comments {
+			if strings.Contains(comment, "Start of tier") {
+				return i
+			}
+		}
+	}
+	return -1
+}
+
+func genInsns(match string) asm.Insns {
```

Comment:
Glad you thought to make this code common! 

This works just fine but I think I might have written it a little differently so that the rules are visible right in the tests rather than being hidden in this function.  Something like this:
```
func genInsns(rule proto.Rule) asm.Insns {
	alloc := idalloc.New()
	setID := func(id string) string {
		alloc.GetOrAlloc(id)
		return id
	}
	pg := NewBuilder(alloc, 1, 2, 3, true)
	rule.Action = "Allow"
	rule.IpVersion = 4
	pol := Rules{
		Tiers: []Tier{{
			Policies: []Policy{{
				Rules: []Rule{{
					Rule: &rule,
				}},
			}},
		}},
	}
	insns, err := pg.Instructions(pol)
	Expect(err).NotTo(HaveOccurred())
	return insns
```
You could also combine `checkLabelsAndComments` and  `genInsns` to make your table even more concise:
```
proto6 := proto.Protocol{NumberOrName: &proto.Protocol_Number{Number: 6}
checkRuleComment(proto.Rule{Protocol: &proto6}, 10, "policy", "label", 0)
```